[IPInt] Add support for the exceptions proposal (except rethrow)

kmiller68 opened the pull request:
#### 1b97cb2b6e051d2f0ff0460714103eb4791a1832
<pre>
[IPInt] Add support for the exceptions proposal (except rethrow)
<a href="https://bugs.webkit.org/show_bug.cgi?id=262446">https://bugs.webkit.org/show_bug.cgi?id=262446</a>

Reviewed by Justin Michaud.

This patch adds support for the Wasm exceptions proposal, with the exception (pun intended) of rethrow.
Rethrow is suffiecently complicated in the IPInt that it should be in its own patch.

Exceptions work mostly as they do in the rest of JSC. When an exception is thrown we write the
target PC and the (newly added) MC to the VM. These tell us where to go once we&apos;ve reentered the
IPInt. There&apos;s also two new entrypoints to the IPInt, one for regular catch blocks and one
for catch_all blocks.

Like the Wasm LLInt, IPInt exception handling works based on PC ranges. The current PC is (now) stored
to the CallFrame&apos;s CallSiteIndex when making calls / performing operations. When an exception is thrown
we check that the PC from the CallFrame is inside the range from one of the function&apos;s HandlerInfo.

This patch also has a couple of other fixes/improvements:
    1) offlineasm self moves should actually be elided. Previously we were checking that the
       two ruby objects had the same address, which didn&apos;t work.
    2) Add a dump to the IPInt, which tells you what the PC and MC should be for each instruction.
    3) Move the _ipint_call_impl call label below the call interpreter macros so it&apos;s clear
       _ipint_call_impl can fall through to .ipint_call_common.
    4) Add a message to the IPInt validation so you know which opcodes are too big.
    5) Rename WasmCodeBlock to UnboxedWasmCalleeStackSlot.
    6) Rename LLIntReturnPC to CallSiteIndex to match CallFrame.

* JSTests/wasm/stress/exception-liveness-tier-up.js:
(assert.eq.): Deleted.
(assert.eq): Deleted.
* JSTests/wasm/stress/exception-simple-throw-catch.js:
(testSimpleTryCatchValue2):
(testNestedCatch):
* JSTests/wasm/stress/exception-throw-from-function-returning-tuple.js:
(testCatchWithExceptionThrownFromFunctionReturningTuple3):
(testCatchWithExceptionThrownFromJSReturningTuple.callback):
(testCatchWithExceptionThrownFromJSReturningTuple):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::CatchInfo::CatchInfo):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/jit/JITExceptions.cpp:
(JSC::genericUnwind):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter.cpp:
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::getStackHeightInValues const):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.cpp:
(JSC::Wasm::HandlerInfo::initialize):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.h:
(JSC::Wasm::UnlinkedHandlerInfo::UnlinkedHandlerInfo):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::condenseControlFlowInstructions):
(JSC::Wasm::IPIntGenerator::addTry):
(JSC::Wasm::IPIntGenerator::convertTryToCatch):
(JSC::Wasm::IPIntGenerator::addCatch):
(JSC::Wasm::IPIntGenerator::addCatchToUnreachable):
(JSC::Wasm::IPIntGenerator::addCatchAllToUnreachable):
(JSC::Wasm::IPIntGenerator::addDelegate):
(JSC::Wasm::IPIntGenerator::addDelegateToUnreachable):
(JSC::Wasm::IPIntGenerator::addThrow):
(JSC::Wasm::IPIntGenerator::addEndToUnreachable):
(JSC::Wasm::IPIntGenerator::dump):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipperInlines.h: Removed.

Canonical link: <a href="https://commits.webkit.org/268872@main">https://commits.webkit.org/268872@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96a48e7889fe4de1f9025c87e117e516bee7ba45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20916 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21321 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22806 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24563 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21501 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21138 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20904 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18162 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23662 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18067 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18976 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25267 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/18227 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19151 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19168 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23179 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/20346 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/19737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/24326 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18986 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5768 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5016 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/25588 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19558 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5593 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
```diff
@@ -287,6 +288,62 @@ WASM_IPINT_EXTERN_CPP_DECL(epilogue_osr, CallFrame* callFrame)
+
+    ASSERT(tag.type().returnsVoid());
+    if (tag.type().numVectors()) {
+        // Note: the spec is still in flux on what to do here, so we conservatively just disallow throwing any vectors.
```
The spec is pretty clear that they can be thrown, we just haven't implemented it yet

kmiller68 replied:
Yeah, I just copied the comment from LLInt.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
```diff
@@ -287,6 +288,62 @@ WASM_IPINT_EXTERN_CPP_DECL(epilogue_osr, CallFrame* callFrame)
+    size_t size = tag.parameterCount();
+    FixedVector<uint64_t> values(size);
+    for (unsigned i = 0; i < size; ++i)
+        values[i] = stack[size - i - 1].u64x2[1];
```
We don't need this in this patch, but we really should add a scalar option to this union

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
```diff
@@ -287,6 +288,62 @@ WASM_IPINT_EXTERN_CPP_DECL(epilogue_osr, CallFrame* callFrame)
+        uint64_t size = wasmException->payload().size();
+
+        for (unsigned i = 0; i < size; ++i)
+            stackPointer[size - i - 1] = { wasmException->payload()[i], wasmException->payload()[i] };
```
Can one of these be 0 or BADBEEF?

kmiller68 replied:
I'll make it 0 in release and 0x1BADBEEF in debug

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -1216,6 +1324,9 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addEndToUnreachable(ControlEntr
     auto block = entry.controlData;
 
+    if (ControlType::isTry(block) || ControlType::isAnyCatch(block))
+        --m_tryDepth;
```
Hmm, no, I am wrong. We delegate to the label of the try, which should jump to the associated catch. So the catch and try need to be at the same depth.

justinmichaud commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -5549,6 +5633,24 @@ elsif X86_64
 end
 
+_ipint_call_impl:
+    # 0 - 3: function index
```
Side comment: we really need a better way to specify these offsets

kmiller68 replied:
Yeah, IPInt has the old LLInt problem that offsets are hardcoded instead of derived from LLIntOffsetsExtractor

justinmichaud commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -933,10 +1017,16 @@ instructionLabel(_return)
 
 instructionLabel(_call)
+    # First store the PC to the call site index
+    storei PC, ArgumentCountIncludingThis + TagOffset[cfr]
```
Should we make a named macro for this instead of the comment?

kmiller68 replied:
Done.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -1153,13 +1204,70 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addCatchAll(Stack&, ControlType
+    m_metadata->m_exceptionHandlers.append({
+        HandlerType::Delegate,
+        static_cast<uint32_t>(data.m_pc),
+        static_cast<uint32_t>(data.m_pcEnd),
```
What does this do if we delegate to a catch-less try? It looks to me like we only ever set endPC in the catch, right?

kmiller68 replied:
I think we just won't find a target handler and will end up propagating the exception out of the function. I'm pretty sure this happens in all our tiers since we always add the HandlerInfo in the addCatch generator callback. This behavior seems weird to me as I would expect to catch the exception at an outer try if one exists. I wonder what other browsers do...

justinmichaud reviewed (approved):
r=me

kmiller68 committed:
[IPInt] Add support for the exceptions proposal (except rethrow)
https://bugs.webkit.org/show_bug.cgi?id=262446

Reviewed by Justin Michaud.

This patch adds support for the Wasm exceptions proposal, with the exception (pun intended) of rethrow.
Rethrow is suffiecently complicated in the IPInt that it should be in its own patch.

Exceptions work mostly as they do in the rest of JSC. When an exception is thrown we write the
target PC and the (newly added) MC to the VM. These tell us where to go once we've reentered the
IPInt. There's also two new entrypoints to the IPInt, one for regular catch blocks and one
for catch_all blocks.

Like the Wasm LLInt, IPInt exception handling works based on PC ranges. The current PC is (now) stored
to the CallFrame's CallSiteIndex when making calls / performing operations. When an exception is thrown
we check that the PC from the CallFrame is inside the range from one of the function's HandlerInfo.

This patch also has a couple of other fixes/improvements:
    1) offlineasm self moves should actually be elided. Previously we were checking that the
       two ruby objects had the same address, which didn't work.
    2) Add a dump to the IPInt, which tells you what the PC and MC should be for each instruction.
    3) Move the _ipint_call_impl call label below the call interpreter macros so it's clear
       _ipint_call_impl can fall through to .ipint_call_common.
    4) Add a message to the IPInt validation so you know which opcodes are too big.
    5) Rename WasmCodeBlock to UnboxedWasmCalleeStackSlot.
    6) Rename LLIntReturnPC to CallSiteIndex to match CallFrame.

* JSTests/wasm/stress/exception-liveness-tier-up.js:
(assert.eq.): Deleted.
(assert.eq): Deleted.
* JSTests/wasm/stress/exception-simple-throw-catch.js:
(testSimpleTryCatchValue2):
(testNestedCatch):
* JSTests/wasm/stress/exception-throw-from-function-returning-tuple.js:
(testCatchWithExceptionThrownFromFunctionReturningTuple3):
(testCatchWithExceptionThrownFromJSReturningTuple.callback):
(testCatchWithExceptionThrownFromJSReturningTuple):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::CatchInfo::CatchInfo):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/jit/JITExceptions.cpp:
(JSC::genericUnwind):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter.cpp:
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::getStackHeightInValues const):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.cpp:
(JSC::Wasm::HandlerInfo::initialize):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.h:
(JSC::Wasm::UnlinkedHandlerInfo::UnlinkedHandlerInfo):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::condenseControlFlowInstructions):
(JSC::Wasm::IPIntGenerator::addTry):
(JSC::Wasm::IPIntGenerator::convertTryToCatch):
(JSC::Wasm::IPIntGenerator::addCatch):
(JSC::Wasm::IPIntGenerator::addCatchToUnreachable):
(JSC::Wasm::IPIntGenerator::addCatchAllToUnreachable):
(JSC::Wasm::IPIntGenerator::addDelegate):
(JSC::Wasm::IPIntGenerator::addDelegateToUnreachable):
(JSC::Wasm::IPIntGenerator::addThrow):
(JSC::Wasm::IPIntGenerator::addEndToUnreachable):
(JSC::Wasm::IPIntGenerator::dump):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipperInlines.h: Removed.

Canonical link: https://commits.webkit.org/268872@main
